### ISSUE TITLE
Feature request: SubImage::copy()

### ISSUE THREAD

**Issue by mark-summerfield** [PistonDevelopers/image#762]
The SubImage::new() function provides a view and so must take a mutable reference to the image.
However, I think it would be useful (certainly for my use cases) to be able to *copy* a sub image. So I'd like a function called, say SubImage::copy() which has the same API as SubImage::new() except that it takes a non-mutable image and returns a fresh image.

**Comment by HeroicKatora** [comment]
It would also be possible to have a `decouple` on SubImage, producing an image with independent lifetime.

**Comment by HeroicKatora** [comment]
As I just noted, `to_owned` already exists, as `to_image`. With the restrictive requirement of `SubImage::new` requiring `&mut I`, it might be best to have `copy` (or similar) on `GenericImage` instead of on `SubImage`. Otherwise, we would confuse the functionality of providing a view into an arbitrary, generic buffer with the functionality of cloning an image into a specific kind of predetermined buffer, `ImageBuffer<I::Pixel, Vec<<I::Pixel as Pixel>::Subpixel>>` as it exists in `SubImage::to_image`.

**Comment by mark-summerfield** [comment]
I think I made a mistake in my original feature request. I don't need to copy the image data, I just need a read-only view on the data and to have the image as non-mut so that I can use a rayon .par_iter() to loop over a bunch of rectangles and examine each one inside the loop using, say, SubImage::read_only(&img, &rect).

**Comment by HeroicKatora** [comment]
In the pull request, `SubImage` has been changed to accept generic reference types, any type implementing `Deref`to be exact, to work around the requirement for `&mut _`. The implementation of the `GenericImage` trait is only required when needing `SubImage` to be a `GenericImage`.

While this opens up many new possibilities, the current state of the rework might be a breaking change of the interface because I'm unsure if all previously possible instantiations still compile.

**Comment by HeroicKatora** [comment]
The interface should be reverse compatible now, `SubImage::new(&mut image)` should now offer the exact same methods and trait implementations as before. Can I request a review of #764 @bvssvni ?

**Comment by HeroicKatora** [comment]
This should be fixed as of #764 .

```
let image = …;
let subimage = image.view(x, y, w, h).to_image();
```
